src/odemis/acq/align/fastem.py

```diff
@@ -18,22 +20,245 @@
+
+    # Connect the future to the task and run it in a thread.
+    # task.run is executed by the executor and runs as soon as no other task is executed
+    _executor.submitf(f, task.run)
+
     return f
 
 
-def _run_fake_alignment():
-    time.sleep(2)
+def estimate_calibration_time(calibrations):
+    """
+    Computes the approximate time it will take to run all calibrations.
+    :param calibrations: (list of str) List of calibrations that should be run.
+    :return (0 <= float): The estimated time for the requested calibrations in s.
+    """
+    tot_time = 0
+    for calib in calibrations:
+        tot_time += calib.estimate_calibration_time()
+
+    return tot_time
+
+
+class CalibrationTask(object):
+    """
+    The calibration task, which runs the calibrations according to the order in the list of calibrations passed.
+    """
+
+    def __init__(self, scanner, multibeam, descanner, detector, stage, ccd, beamshift, det_rotator, future, calibrations):
+        """
+        :param scanner: (xt_client.Scanner) Scanner component connecting to the XT adapter.
+        :param multibeam: (technolution.EBeamScanner) The multibeam scanner component of the acquisition server module.
+        :param descanner: (technolution.MirrorDescanner) The mirror descanner component of the acquisition server module.
+        :param detector: (technolution.MPPC) The detector object to be used for collecting the image data.
+        :param stage: (actuator.ConvertStage) The stage in the corrected scan coordinate system, the x and y axes are
+            aligned with the x and y axes of the multiprobe and the multibeam scanner.
+        :param ccd: (model.DigitalCamera) A camera object of the diagnostic camera.
+        :param beamshift: (tfsbc.BeamShiftController) Component that controls the beamshift deflection.
+        :param det_rotator: (tmcm.CANController) K-mirror controller.
+        :param calibrations: (list of str) List of calibrations that should be run.
+
+        :param future: (ProgressiveFuture) Acquisition future object, which can be cancelled.
+                        (Exception or None): Exception raised during the calibration or None.
+        """
+        self._scanner = scanner
+        self._multibeam = multibeam
+        self._descanner = descanner
+        self._detector = detector
+        self._dataflow = detector.data
+        self._stage = stage
+        self._ccd = ccd
+        self._beamshift = beamshift
+        self._det_rotator = det_rotator
+        self._future = future
+
+        self.calibrations = calibrations
+
+        # List of calibrations to be executed. Used for progress update.
+        self._calibrations_remaining = set(calibrations)
+
+        # keep track if future was cancelled or not
+        self._cancelled = False
+
+        if model.MD_CALIB not in self._beamshift.getMetadata().keys():
+            self._beamshift.updateMetadata({model.MD_CALIB: self._scanner.beamShiftTransformationMatrix.value})
```

Comment:
Can you add a comment why this is necessary here, or will this be removed in another PR anyway?

Also just fyi, the `.keys()` is not necessary `if model.MD_CALIB not in self._beamshift.getMetadata():` also works, but use whatever you feel is more readable :)

Reply:
Removed now as we have a separate pr that loads this MD at init of the beamshift component. #pr2031